Add `ProductIcon` component to `@automattic/components` package

delawski opened the pull request:
As proposed in pbtFFM-11-p2, this PR updates and moves the `ProductIcon` component to the `@automattic/components` package.

<img width="799" alt="Screenshot 2019-12-06 at 18 33 59" src="https://user-images.githubusercontent.com/478735/70343219-00a11700-1857-11ea-930c-b0ef1e94ec39.png">

Note that this PR doesn't replace the existing occurrences of the `ProductIcon` that can be found in the `client/components` folder. The adoption of the new component in Calypso (and Jetpack) will be part of a subsequent PRs.

#### Changes proposed in this Pull Request

* Add `ProductIcon` component to the `@automattic/components` package

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to https://wpcalypso.wordpress.com/devdocs/design/product-icon
* Confirm that the plans and products icons are rendered correctly.

Fixes n/a

Commit:
Add `ProductIcon` component to `@automattic/components` package

keoshi reviewed (approved):
Looking great to me!

tyxla commented on packages/components/src/product-icon/README.md:
```diff
@@ -0,0 +1,45 @@
+ProductIcon Component
```
I think it would be useful that if in this README we have a step-by-step reference of how to add a new icon if we want to.

tyxla commented on packages/components/src/product-icon/index.jsx:
```diff
@@ -0,0 +1,39 @@
+function ProductIcon( { className, slug } ) {
+	return (
+		<img
+			src={ paths[ slug ] }
```
Should we render nothing instead of an empty `img` when we've specified an unsupported `slug`?

delawski replied:
Wouldn't the `PropTypes` like this:
```js
slug: PropTypes.oneOf( Object.keys( paths ) ).isRequired,
```
prevent the component from getting an unsupported `slug`?

tyxla replied:
No. You'll get a warning that that slug is unexpected, but component will still get the unsupported props and render unexpected results.

delawski replied:
I'm going to add a check in this case 👍

blowery replied:
Could we not use slugs at all? Instead, export individual components for each product. That way, if a consumer only needs one of the products, they don't have to import everything and pay the cost of all packages. For consumers that want a slug-to-icon mapping, we could offer a factory function that does that separately.

blowery replied:
Hrm, though individual exports would make the main interface on `@automattic/components` a bit messy. `import { JetpackFreeProductIcon } from '@automattic/components'` feels a bit too verbose.

I think this is fine to land as-is. Maybe it makes more sense for this to live in its own package though? Releasing point releases to `@automattic/components` when we release new plan types feels odd.

tyxla replied:
> Maybe it makes more sense for this to live in its own package though? Releasing point releases to @automattic/components when we release new plan types feels odd.

I don't see why it's odd TBH - we made a conscious decision to make this a reusable component and it makes a lot of sense in the a8c world, because we need it in multiple places. In that case, it only makes sense to release a new version when we add a new feature (icon) to a component (`ProductIcon`). 

If this sentiment is purely about point release vs major release, I agree, it would make sense to release a major release when including a new icon, and not prefer a point release.

tyxla commented on packages/components/src/product-icon/paths.jsx:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Internal dependencies
+ */
+import free from './images/free.svg';
```
So right now, we'll duplicate `./images/` for each icon, and I don't see a good reason for that.

I wonder if we could reduce some of the overhead:

* Defining an `icons` array (here or even as a `const` inline with the `ProductIcon` component) with all the expected icons, no need to be an object and duplicate .
* Having one less object AND file. 
* Adding `./images/` to the beginning of the `img src`, and adding the `slug` prop afterwards, to avoid the need for specifying the paths here.
*  Simplifying `PropTypes.oneOf( Object.keys( paths ) ).isRequired,` to `PropTypes.oneOf( icons ).isRequired,`.

delawski replied:
> Adding `./images/` to the beginning of the `img src`, and adding the `slug` prop afterwards, to avoid the need for specifying the paths here.

I think it won't work here. We need a full path to an image in the `import` rule so that Webpack can transform this path to the actual, physical path, where the picture is located. If we remove the `./image` part, Webpack won't be able to build up the proper path.

tyxla replied:
Good point, I think I overlooked that. Disregard 👍

jsnajdr replied:
Some webpack magic like this could work:
```js
const imagePath = await import(
  /* webpackInclude: /\.svg$/ */
  /* webpackMode: "eager" */
  `./images/${ icon }`
);
```
That allows us to construct the import path dynamically and give webpack hints at compile time about what the valid values of `icon` can be.

`webpackMode: eager` instructs webpack to not create any extra chunks for the dynamic imports.

The disadvantage of this is that the `@automattic/components` library now needs webpack to work, and that the `import()` is dynamic -- it returns a promise.

The webpack-specific `require.context()` function could solve the second issue, but the first one remains.

I'm fine with the existing explicit and somewhat verbose solution, as it involves zero magic 🙂

tyxla replied:
Yeah, exactly, I think we should try to keep it clean and with zero magic when possible 😉

blowery replied:
@jsnajdr isn't this using the file-loader? It just returns paths, not SVG components.

jsnajdr replied:
> isn't this using the file-loader? It just returns paths, not SVG components.

Yes it is! The only thing we'd be able to tree shake are the path strings, which is not a great win. I didn't realize that.

tyxla reviewed (commented):
This is working great @delawski 😉 Stellar work 🙌 

I think there are little things that can be optimized, but I'd like to discuss them anyway. I have left some questions and comments. Let me know what you think.

jsnajdr commented on packages/components/src/product-icon/index.jsx:
```diff
@@ -0,0 +1,39 @@
+			className={ classNames(
+				'product-icon',
+				`product-icon__${ slug }`,
+				`is-${ slug }`,
```
The `product-icon__${ slug }` and `is-${ slug }` seem to be redundant -- they serve the same need (customize by specific icon) and we need only one.

Can we clean this up in the process of moving the component to `@automattic/components`?

delawski replied:
Sounds perfect! 

Originally, in the `PlanIcon` component, the modifier was in a form of `product-icon__${ slug }`. I left for backwards compatibility. I haven't actually checked if it's used at all, though.

I will go through the codebase and replace any occurrences with a `is-${ slug }` (which seems to be a more up-to-date pattern in terms of our coding standards).

jsnajdr commented on packages/components/src/product-icon/README.md:
````diff
@@ -0,0 +1,45 @@
+## Usage
+
+```jsx
+import ProductIcon from '@automattic/components/product-icon';
````
Yes, let's use `import { ProductIcon } from '@automattic/components'`. It's inline with existing convention in `@automattic/components` and `@wordpress/components`.

And `import from '@automattic/components/product-icon'` won't even work I'm afraid. The package's `package.json` specifies `main` and `module` fields that point to `dist/cjm` and `dist/esm` build directories. With the import in form `.../product-icon`, these fields won't be used and the bundler will fail to find the transpiled module.

We'd need a per-component `package.json` to enable the subdirectory imports.

Commit:
Address code review feedback.

delawski commented:
@tyxla @jsnajdr I've addressed your feedback in [the latest commit](https://github.com/Automattic/wp-calypso/pull/38253/commits/b9e9c409f79aa3f2e509c5ac2e625f4ae7bc7f12). Please have another look.

tyxla commented on packages/components/src/product-icon/README.md:
```diff
@@ -0,0 +1,57 @@
+In order to add a new icon to the component, do the following:
+* Add a new icon image to the `images` folder.
+* Import the image in the `paths.js` file. This way an image URL will be generated by a bundler for further reuse.
+* Add the image path to the default export block in the `paths.js` file.
```
Another point to add here: `Document the new icon as an available option for the `slug` prop in this README.`

delawski replied:
I was thinking of this one and totally forgot when elaborating on the assets imports. I'll update the readme 👍

tyxla commented on packages/components/src/product-icon/README.md:
```diff
@@ -0,0 +1,57 @@
+In order to add a new icon to the component, do the following:
+* Add a new icon image to the `images` folder.
+* Import the image in the `paths.js` file. This way an image URL will be generated by a bundler for further reuse.
+* Add the image path to the default export block in the `paths.js` file.
```
And another point you could consider adding here: `Optionally, update the docs/example.jsx file to include the new icon.`. That would be optional because one might want to not display that icon in the example sometimes.

tyxla reviewed (approved):
This LGTM 🚢 Nice work!

I've left a couple of non-blocking suggestions for the README, but other than that, this is ready to ship!

tyxla reviewed (changes requested):
Actually, something I've noticed is we have missed to import the component in `packages/components/src/index.js` to make it usable. Is that intentionally left out?

Commit:
Address second round of feedback.

delawski commented:
@tyxla @jsnajdr I've [updated the `index.js` in the `packages/components`](https://github.com/Automattic/wp-calypso/pull/38253/commits/a2676af65b7915547150081376cdf4c90e78523e) so that now it includes the `ProductIcon` export. This however broke the build process and now I'm getting such errors for each icon image:

```node
ERROR in ./packages/components/dist/esm/product-icon/paths.js
Module not found: Error: Can't resolve '../images/premium.svg' in '/Users/piotr/a8c/calypso/packages/components/dist/esm/product-icon'
 @ ./packages/components/dist/esm/product-icon/paths.js 16:38-70
 @ ./packages/components/dist/esm/product-icon/index.js
 @ ./packages/components/dist/esm/index.js
 @ ./client/components/button/index.js
 @ ./client/login/magic-login/handle-emailed-link-form.jsx
 @ ./client/login/controller.js
 @ ./client/login/index.node.js
 @ ./server/pages/index.js
 @ ./server/boot/index.js
 @ ./index.js
```

Any ideas what might be wrong?

jsnajdr commented:
> Any ideas what might be wrong?

That's because the dist directories with transpiled files, `dist/esm` and `dist/cjs`, don't contain the SVG assets. They just were not copied there.

For SCSS styles, we have the `copy-styles` command in `calypso-build` that does copy the style files and is invoked in the `npm prepare` script of `package/components`.

We need to add a similar `copy-assets` script. Or rename the `copy-styles` and make it more generic. It's really about just copying files with certain extensions. What does @ockham think?

delawski commented:
@jsnajdr Thank you for having a look!

> We need to add a similar `copy-assets` script. Or rename the `copy-styles` and make it more generic.

I tried both approaches locally and I'm getting another kind of error in both cases:

```node
@automattic/components: sh: copy-assets: command not found
```

It happens even though I've replaced all occurrences of `copy-styles` with `copy-assets` (or added a new script like this).

I've updated `packages/calypso-build/package.json` file like:

```diff
 	"bin": {
 		"transpile": "./bin/transpile.js",
 		"calypso-build": "./bin/calypso-build.js",
-		"copy-styles": "./bin/copy-styles.js"
+		"copy-styles": "./bin/copy-styles.js",
+		"copy-assets": "./bin/copy-assets.js"
 	},
```

and `components/package.json` file like:

```diff
 	"scripts": {
 		"clean": "npx rimraf dist types",
 		"prepublish": "npm run clean",
-		"prepare": "transpile && tsc && copy-styles"
+		"prepare": "transpile && tsc && copy-styles && copy-assets"
 	}
```

And of course I've added the `packages/calypso-build/bin/copy-assets.js` file itself with such a change compared to `packages/calypso-build/bin/copy-styles.js`:

```diff
 const copyOptions = {
 	overwrite: true,
-	filter: [ '**/*.scss', '!**/test/**' ],
+	filter: [ '**/*.gif', '**/*.jpg', '**/*.jpeg', '**/*.png', '**/*.svg' ],
 	concurrency: 127,
 };
```

I'm not too familiar with Calypso build tooling and it's hard for me to troubleshoot. Any ideas what am I missing?

jsnajdr commented:
@delawski The `copy-assets` changes you describe in your comment look correct to me. It should work. Can you verify that you have the following symlinks in your root `node_modules/.bin` directory?
```
node_modules/.bin/copy-assets -> ../@automattic/calypso-build/bin/copy-assets.js
node_modules/.bin/copy-styles -> ../@automattic/calypso-build/bin/copy-styles.js
```
You might need to re-run `npm i` or `npm ci` to force NPM to create/update these symlinks.

You can also push your unfinished changes to this branch, so that others can debug them, too.

jsnajdr commented on packages/components/src/product-icon/README.md:
```diff
@@ -0,0 +1,59 @@
+- `personal`
+- `premium`
+- `business`
+- `ecommerce`
```
There's one more nice-to-have thing: can we prefix these WP.com plans with `wpcom-*`? That would put WP.com and Jetpack on the same level of hierarchy:
```
wpcom-free
wpcom-personal
jetpack-free
jetpack-personal
```

If that breaks all existing references and constants in Calypso and would require some mapping code, then it's not worth it, though.

delawski replied:
This suggestion makes sense. I've added the prefix in [the latest commit](https://github.com/Automattic/wp-calypso/pull/38253/commits/4254a82ff5e5dcda1084d4828344c88e8346d380).

When it comes to the existing references, we will have to update and adjust them in any case since the existing constants values in Calypso are quite messed up. E.g. for the Free plans we have `free_plan` and `jetpack_free` while Personal is `personal-bundle` and `jetpack_personal`.

Commit:
Add support for assets processing inside the packages

Commit:
Use a prefix in WPCOM plans slugs and icons names

delawski commented:
@jsnajdr Thank you for the support! Indeed, I was missing a symlink to `copy-assets.js` inside `node_modules/.bin`. After running `npm ci` everything started to work as expected.

I've pushed my [updates to the `calypso-build` in the recent commit](https://github.com/Automattic/wp-calypso/pull/38253/commits/4dc1acbd29d4160753155bd00813452153662242). They are pretty much the same as I described in the previous comment.

@tyxla @jsnajdr Can you give this PR a try once again so that it can be hopefully merged soon?

tyxla reviewed (approved):
This looks great and tests well 👍 

Ideally, the changes to the calypso-build package would be their own PR, but I guess that's not super critical.

Feels ready to 🚢 from my perspective.

tyxla commented:
cc @Automattic/team-calypso and @ockham - this change will need a release of `calypso-build`.

jsnajdr commented:
One more thing: let's add a line to `calypso-build/CHANGELOG.md` about the new `copy-assets` script.

jsnajdr commented:
> this change will need a release of `calypso-build`.

It's going to be 5.1.0, right? No breaking changes since 5.0.1, adds new functionality.

delawski commented:
@jsnajdr Should I add an entry in the `components` package as well?

Also, should I bump up the minor versions in both packages?

jsnajdr commented:
> Should I add an entry in the `components` package as well?

Yes please 🙂 

> Also, should I bump up the minor versions in both packages?

I'm not sure -- maybe it should be done later, as an independent commit at the time of publishing, when going through the [publishing workflow](https://github.com/Automattic/wp-calypso/blob/master/docs/monorepo.md#publishing)? @sirreal and @ockham are the authorities on this 🙂

Commit:
Update changelogs

delawski commented:
[The changelogs are now updated](https://github.com/Automattic/wp-calypso/pull/38253/commits/589dc4510079c2685f3cdbab9768a7a498e9874c).

I guess the PR is ready for deployment now?

tyxla commented:
It is ready from my perspective. @jsnajdr does it look good for you?

jsnajdr reviewed (approved):
I think this is shippable :+1: @blowery shared some good points about making the component better optimizable. We can address these in this PR or a followup -- it's up to you. Just let's not publish the new package version right away, so that we don't find ourselves inside a semver headache if we break the API.

delawski commented:
@jsnajdr Thanks for the confirmation 🙂

Btw. where can I find @blowery suggestions? Maybe it's worth taking care of the issues as part of this PR?

jsnajdr commented:
@delawski I was referring to this comment: https://github.com/Automattic/wp-calypso/pull/38253#discussion_r356027572

With the component as is, a minimalistic React app:
```js
import React from 'react';
import ReactDom from 'react-dom';
import { ProductIcon } from '@automattic/components';

ReactDom.render( <ProductIcon slug="wpcom-free" />, document.body );
```
will bundle not only the SVG image for `wpcom-free`, but for all other product icons as well. In ideal case, the code and assets would be optimized to include only the one SVG image we actually use.

blowery commented:
@delawski @jsnajdr after mulling it over more, I think this is fine to land as-is. Down the road, we may want to move this kind of thing to its own package though, like `@automattic/products` or something.

tyxla commented:
I think that what @jsnajdr suggested is a nice optimization we could address in another PR. @delawski what do you think?

And on `@automattic/products` - I don't see a lot of value in having this its own package, but I might be missing some context 😉

ockham commented:
Thanks a lot, @delawski, this is stellar work! :star2: 

> We need to add a similar `copy-assets` script. Or rename the `copy-styles` and make it more generic. It's really about just copying files with certain extensions. What does @ockham think?

I think one shared `copy-assets` script that's in charge of copying both styles and other assets makes a lot of sense. We can address that in a follow-up PR though in order to not further delay this one.

> It's going to be 5.1.0, right? No breaking changes since 5.0.1, adds new functionality.

Seems right :+1: 

> I'm not sure -- maybe it should be done later, as an independent commit at the time of publishing, when going through the [publishing workflow](https://github.com/Automattic/wp-calypso/blob/master/docs/monorepo.md#publishing)? @sirreal and @ockham are the authorities on this 🙂 

Probably makes sense to do bump versions only right before publishing, yeah.

Commit:
Replace copy-styles with generic copy-assets to handle styles and images

delawski commented:
> @delawski @jsnajdr after mulling it over more, I think this is fine to land as-is.
> I think that what @jsnajdr suggested is a nice optimization we could address in another PR. @delawski what do you think?

@blowery @tyxla I can see there's an agreement on this point. My opinion is the same - let's not optimize now and leave it for another iteration.

> I think one shared copy-assets script that's in charge of copying both styles and other assets makes a lot of sense.

@ockham Thank you for chiming in! This makes a lot of sense to me as well. Since it's not a huge change in the code, I decided to give it a go in [the latest commit](https://github.com/Automattic/wp-calypso/pull/38253/commits/830419e0eff45e53ce27652e80fd10ef5f0736e5). Please have a look if you get a chance 🙂

jsnajdr commented:
> This makes a lot of sense to me as well. Since it's not a huge change in the code, I decided to give it a go in the latest commit.

I think it's a good decision to replace `copy-styles` with generic `copy-assets`. Just keep in mind that it makes the change technically a major semver update. It's a breaking change.

delawski commented:
> Just keep in mind that it makes the change technically a major semver update. It's a breaking change.

@jsnajdr in this case do we want to go down this road? I haven't yet been involved in such a change. What are the drawbacks and potential risks if we introduce such a breaking change?

tyxla reviewed (approved):
I'm quite happy with where this went, and see no blockers from my side. LGTM 👍 

Would love to see a 👍 👎 from @Automattic/team-calypso before shipping though, and specifically by @jsnajdr @blowery @ockham.

jsnajdr commented:
I already approved it and my :+1: still stands :shipit:

jsnajdr commented:
> in this case do we want to go down this road? I haven't yet been involved in such a change. What are the drawbacks and potential risks if we introduce such a breaking change?

There's no risk. Just a reminder that we'll need to release 6.0.0. With semver, that's fine. There's no marketing or emotional meaning attached to a major version bump. It just says that there are breaking changes and the package consumers will need to do some fixing.
